src/backends/github/API.js

```diff
@@ -507,44 +437,37 @@ export default class API {
         content: contentBase64,
         encoding: "base64",
       }),
-    }).then((response) => {
-      item.sha = response.sha;
-      item.uploaded = true;
-      return item;
+    })).then(response => Object.assign({}, item, {
+      sha: response.sha,
+      uploaded: true,
     }));
   }
 
+  isFile(obj) {
+    return obj.file;
+  }
+
   updateTree(sha, path, fileTree) {
     return this.getTree(sha)
-      .then((tree) => {
-        let obj;
-        let filename;
-        let fileOrDir;
-        const updates = [];
-        const added = {};
-
-        for (let i = 0, len = tree.tree.length; i < len; i++) {
-          obj = tree.tree[i];
-          if (fileOrDir = fileTree[obj.path]) {
-            added[obj.path] = true;
-            if (fileOrDir.file) {
-              updates.push({ path: obj.path, mode: obj.mode, type: obj.type, sha: fileOrDir.sha });
-            } else {
-              updates.push(this.updateTree(obj.sha, obj.path, fileOrDir));
-            }
-          }
-        }
-        for (filename in fileTree) {
-          fileOrDir = fileTree[filename];
-          if (added[filename]) { continue; }
-          updates.push(
-            fileOrDir.file ?
-              { path: filename, mode: "100644", type: "blob", sha: fileOrDir.sha } :
-              this.updateTree(null, filename, fileOrDir)
-          );
-        }
+      .then(({ tree: dirContents }) => {
+        const updatedItems = dirContents.filter(item => fileTree[item.path]);
+        const added = zipObject(updatedItems.map(item => item.path), Array(updatedItems.length).fill(true));
+        const [updatedFiles, updatedDirs] = partition(updatedItems, this.isFile);
+        const updatePromises = [
+          ...updatedDirs.map(dir => this.updateTree(dir.sha, dir.path, fileTree[dir.path])),
+          ...updatedFiles.map(file => ({ ...pick(file, ['path', 'mode', 'type']), sha: fileTree[file.path].sha })),
+        ];
+
+        const newItems = entries(fileTree).filter(([filename]) => !added[filename]);
+        const [newFiles, newDirs] = partition(newItems, ([, file]) => this.isFile(file));
+        const newPromises = [
+          ...newDirs.map(([dirName, dir]) => this.updateTree(null, dirName, dir)),
+          ...newFiles.map(([fileName, file]) => ({ path: fileName, mode: "100644", type: "blob", sha: file.sha })),
+        ];
+
+        const updates = [...updatePromises, ...newPromises];
         return Promise.all(updates)
-          .then(updates => this.request(`${ this.repoURL }/git/trees`, {
+          .then(resolvedUpdates => this.request(`${ this.repoURL }/git/trees`, {
             method: "POST",
             body: JSON.stringify({ base_tree: sha, tree: updates }),
```

Comment:
Shouldn't the `tree` be `resolvedUpdates` here?

I'm was having a problem with `updateTree` where sometimes it will request a tree update, but with tree set to `[{}]` (empty).

Reply:
Great catch, thanks.

Reply:
@Benaiah Just wondering, are you guys planning on getting the tests done before you merge this?

Reply:
This still needs to be fixed.